### PR TITLE
Version 4.72 - speedups to case, RNA alignments and Pydantic2

### PR DESCRIPTION
**Pull request by dnil**
This PR marks a new Scout release. We apply semantic versioning. This is a minor release as it adds and changes significant features in a backwards compatible manner.

### Added
- [x] A GitHub action that checks for broken internal links in docs pages
- [x] Link validation settings in mkdocs.yml file
- [x] Load and display full RNA alignments on alignment viewer
- [x] Genome build check when loading a case
- [ ] Extend event index to previous causative variants and always load them
### Fixed
- [x] Documentation nav links for a few documents
- [x] Slightly extended the BioNano Genomics Access integration docs
- [x] ~~Loading of SVs when VCF is missing the INFO.END field but has INFO.SVLEN field~~ (tested by another facility)
- [x] Escape protein sequence name (if available) in case general report to render special characters correctly
- [x] CaseS HPO term searches for multiple terms works independent of order
- [x] CaseS search regexp should not allow backslash
- [x] CaseS cohort tags can contain whitespace and still match
- [x] Remove diagnoses from cases even if OMIM term is not found in the database
- [x] Parsing of disease-associated genes
- [X] Displaying custom case images loaded with scout version <= 4.71
- [x] Use pydantic version >=2 in requirements.txt file
### Changed
- [x] Column width adjustment on caseS page
- [x] Use Python 3.11 in tests
- [x] Update some github actions
- [x] Upgraded Pydantic to version 2
- [x] Case validation fails on loading when associated files (alignments, VCFs and reports) are not present on disk
- [x] Case validation fails on loading when custom images have format different then ["gif", "svg", "png", "jpg", "jpeg"]
- [x] Custom images keys `case` and `str` in case config yaml file are renamed to `case_images` and `srt_variants_images`
- [x] Simplify and speed up case general report code
- [x] ~~Speed up case retrieval in case_matching_causatives~~ (tested by another facility)
- [x] Stricter validation of mandatory genome build key when loading a case. Allowed values are ['37','38',37,38]
- [x] Improved readability of variants length and coordinates on variantS pages

## Update considerations
- [ ] Remember to update changed indices, in particular `scout index --update -c event`

## Solna only tasks for deployment
- [ ] Update HPO (#4112)
- [ ] Merge updated config for bionano access (https://github.com/Clinical-Genomics/servers/pull/1144)

<details>
<summary>Testing on cg-vm1 server (Clinical Genomics Stockholm)</summary>

**Prepare for testing**
1. Make sure the PR is pushed and available on [Docker Hub](https://hub.docker.com/repository/docker/clinicalgenomics/scout-server-stage)
1. Fist book your testing time using the Pax software available at [https://pax.scilifelab.se/](https://pax.scilifelab.se). The resource you are going to call dibs on is `scout-stage` and the server is `cg-vm1`.
1. `ssh <USER.NAME>@cg-vm1.scilifelab.se`
1. `sudo -iu hiseq.clinical`
1. `ssh localhost`
1. (optional) Find out which scout branch is currently deployed on cg-vm1: `podman ps`
1. Stop the service with current deployed branch: `systemctl --user stop scout.target`
1. Start the scout service with the branch to test: `systemctl --user start scout@<this_branch>`
1. Make sure the branch is deployed: `systemctl --user status scout.target`
1. After testing is done, repeat procedure at [https://pax.scilifelab.se/](https://pax.scilifelab.se), which will release the allocated resource (`scout-stage`) to be used for testing by other users.
</details>

<details>
<summary>Testing on hasta server (Clinical Genomics Stockholm)</summary>

**Prepare for testing**
1. `ssh <USER.NAME>@hasta.scilifelab.se`
1. Book your testing time using the Pax software. `us; paxa -u <user> -s hasta -r scout-stage`. You can also use the WSGI Pax app available at [https://pax.scilifelab.se/](https://pax.scilifelab.se).
1. (optional) Find out which scout branch is currently deployed on cg-vm1: `conda activate S_scout; pip freeze | grep scout-browser`
1. Deploy the branch to test: `bash /home/proj/production/servers/resources/hasta.scilifelab.se/update-tool-stage.sh -e S_scout -t scout -b <this_branch>`
1. Make sure the branch is deployed: `us; scout --version`
1. After testing is done, repeat the `paxa` procedure, which will release the allocated resource (`scout-stage`) to be used for testing by other users.
</details>

**How to test**:
1. how to test it, possibly with real cases/data

**Expected outcome**:
The functionality should be working
Take a screenshot and attach or copy/paste the output.

**Review:**
- [x] code approved by CR
- [x] tests executed by DN, CR

### CODE REVIEW

**Comment by dnil** [comment]
We should wait for https://github.com/Clinical-Genomics/scout/pull/4114 as well!

**Comment by dnil** [comment]
Maybe we should add a small blog entry for this release, in particular for the index update, as well as the load config file changes to make it obvious to other centers?

**Comment by northwestwitch** [comment]
> Maybe we should add a small blog entry for this release, in particular for the index update, as well as the load config file changes to make it obvious to other centers?

Yes, sure, could you do that? I feel you have more control, especially for the bionano thing

**Comment by dnil** [comment]
The bionano is a local, just a credentials change really. I can say something about the index if you write about the config file change!

**Comment by northwestwitch** [comment]
While writing on the the blog page it came to my mind that sometimes a new cases is loaded and it's missing the delivery report (which is created at a later stage perhaps). I wonder if I've introduced changes in the case validation that will prevent this behaviour.  In that case I think I should remove that check. I'll check!

**Comment by northwestwitch** [comment]
> While writing on the the blog page it came to my mind that sometimes a new case is loaded and it's missing the delivery report (which is created at a later stage perhaps). I wonder if I've introduced changes in the case validation that will prevent this behaviour. In that case I think I should remove that check. I'll check

Yes, that is the case, being these the lists of items that get check for path validity:

```
SAMPLES_FILE_PATH_CHECKS = [
    "bam_file",
    "mitodel_file",
    "rhocall_bed",
    "rhocall_wig",
    "rna_alignment_path",
    "rna_coverage_bigwig",
    "splice_junctions_bed",
    "tiddit_coverage_wig",
    "upd_regions_bed",
    "upd_sites_bed",
    "vcf2cytosure",
]

CASE_FILE_PATH_CHECKS = [
    "cnv_report",
    "coverage_qc_report",
    "delivery_report",
    "gene_fusion_report",
    "gene_fusion_report_research",
    "madeline_info",
    "multiqc",
    "multiqc_rna",
    "peddy_ped",
    "peddy_ped_check",
    "peddy_sex_check",
    "RNAfusion_inspector",
    "RNAfusion_report",
    "RNAfusion_report_research",
]

VCF_FILE_PATH_CHECKS = [
    "vcf_cancer",
    "vcf_cancer_research",
    "vcf_cancer_sv",
    "vcf_cancer_sv_research",
    "vcf_snv",
    "vcf_snv_research",
    "vcf_mei",
    "vcf_mei_research",
    "vcf_str",
    "vcf_sv",
    "vcf_sv_research",
]

```

I think I will exclude the check for the list of items in the middle (CASE_FILE_PATH_CHECKS). what do ou think @dnil ?

**Comment by dnil** [comment]
> > While writing on the the blog page it came to my mind that sometimes a new case is loaded and it's missing the delivery report (which is created at a later stage perhaps). I wonder if I've introduced changes in the case validation that will prevent this behaviour. In that case I think I should remove that check. I'll check
> 
> Yes, that is the case, being these the lists of items that get checked for path validity:
> 
> ```
> SAMPLES_FILE_PATH_CHECKS = [
>     "bam_file",
>     "mitodel_file",
>     "rhocall_bed",
>     "rhocall_wig",
>     "rna_alignment_path",
>     "rna_coverage_bigwig",
>     "splice_junctions_bed",
>     "tiddit_coverage_wig",
>     "upd_regions_bed",
>     "upd_sites_bed",
>     "vcf2cytosure",
> ]
> 
> CASE_FILE_PATH_CHECKS = [
>     "cnv_report",
>     "coverage_qc_report",
>     "delivery_report",
>     "gene_fusion_report",
>     "gene_fusion_report_research",
>     "madeline_info",
>     "multiqc",
>     "multiqc_rna",
>     "peddy_ped",
>     "peddy_ped_check",
>     "peddy_sex_check",
>     "RNAfusion_inspector",
>     "RNAfusion_report",
>     "RNAfusion_report_research",
> ]
> 
> VCF_FILE_PATH_CHECKS = [
>     "vcf_cancer",
>     "vcf_cancer_research",
>     "vcf_cancer_sv",
>     "vcf_cancer_sv_research",
>     "vcf_snv",
>     "vcf_snv_research",
>     "vcf_mei",
>     "vcf_mei_research",
>     "vcf_str",
>     "vcf_sv",
>     "vcf_sv_research",
> ]
> ```
> 
> I think I will exclude the check for the list of items in the middle (CASE_FILE_PATH_CHECKS). or at least some items from it, what do you think @dnil ?

Could you explain in a little more detail how you view the problem scenario? Afaik we check for the existence of files listed in the load config. It seems reasonable to assume any file mentioned in the load config, but missing from disk is a mistake? Typically when reports are added later, aren't they loaded via a `scout load *-report` CLI mechanism?

**Comment by northwestwitch** [comment]
> Could you explain in a little more detail how you view the problem scenario? Afaik we check for the existence of files listed in the load config. It seems reasonable to assume any file mentioned in the load config, but missing from disk is a mistake? Typically when reports are added later, aren't they loaded via a `scout load *-report` CLI mechanism?

Ok, I've asked prod and looks like we are safe here, since all reports are created before. I just wanted to be sure that nothing would break

**Comment by northwestwitch** [comment]
Sorry I pushed to the wrong branch, I'll revert!

File: scout/build/disease.py [hunk 1]
```diff
@@ -42,7 +42,6 @@ def build_disease_term(disease_info: dict, alias_genes: dict = {}) -> dict:
 
     disease_obj["genes"] = list(hgnc_ids)
 
-    LOG.warning(disease_obj)
```
Comment (dnil) [inline scout/build/disease.py:45]: good stuff though.. 😸 I assume we will have a different PR for it?

Comment (northwestwitch) [reply scout/build/disease.py]: yes!

**Comment by northwestwitch** [comment]
Tested the loading of custom images, on [case](https://scout-stage.scilifelab.se/cust000/F0006358-mip7):

image

And [str image](https://scout-stage.scilifelab.se/cust000/F0006358-mip7/7ed1c05aa886ec700af8b93e26c492f4):
image

**Comment by northwestwitch** [comment]
Test loading a case image and a str variant image using the old config file, from the branch of the old release (version_4.72):

Using the new version for the web server (4.72):

Variant page: 🆗 

image


But case page has no image! ⛔ 

I'll check what's going and fix!

**Comment by northwestwitch** [comment]
Testing loading a case without genome build in config file:

image

Test with non-conforming build:

image

**Comment by northwestwitch** [comment]
> Loaded a case with an RNA alignment path;
> 

Nice!

**Review by northwestwitch — APPROVED** [review]
This release looks great! 💯

**Comment by dnil** [comment]
Marked a low-scoring variant (score -2) causative in 15030-miptest, checked that the variant was not found in 15030-mip7, then re uploaded it and voila:
Screenshot 2023-10-18 at 18 13 43

**Comment by northwestwitch** [comment]
Deployed now on Hasta prod:

image

**Comment by northwestwitch** [comment]
Deployed on cg-prod-services:

image

**Pull request merged**